services/datalad/datalad_service/tasks/publish.py

```diff
@@ -104,48 +109,73 @@ def get_dataset_realm(ds, siblings, realm=None):
+    def should_export(ds, tags):
+        return True
+    export_all_tags(store, dataset, cookies, get_realm, should_export)
+    
+
+def export_all_tags(store, dataset, cookies, get_realm, check_should_export, esLogger=None):
     """Migrate a dataset and all snapshots to an S3 bucket"""
-    realm = get_s3_realm(realm=realm)
     dataset_id = dataset
     ds = store.get_dataset(dataset)
     tags = [tag['name'] for tag in ds.repo.get_tags()]
     siblings = ds.siblings()
-    s3_remote = s3_sibling(ds, siblings, realm=realm)
-    for tag in tags:
-        publish_target(ds, realm.s3_remote, tag)
-        gevent.sleep()
-        # Public publishes to GitHub
-        if realm == DatasetRealm.PUBLIC and DATALAD_GITHUB_EXPORTS_ENABLED:
-            github_remote = github_sibling(ds, dataset_id, siblings)
-            publish_target(ds, realm.github_remote, tag)
-            gevent.sleep()
-    clear_dataset_cache(dataset, cookies)
-
-
-def publish_snapshot(store, dataset, snapshot, cookies=None, realm=None):
-    """Publish a snapshot tag to S3, GitHub or both."""
-    ds = store.get_dataset(dataset)
-    siblings = ds.siblings()
-
-    realm = get_dataset_realm(ds, siblings, realm)
-
-    # Create the sibling if it does not exist
-    s3_sibling(ds, siblings)
-    gevent.sleep()
-
-    # Export to S3
-    publish_target(ds, realm.s3_remote, snapshot)
-    gevent.sleep()
-
-    # Public publishes to GitHub
-    if realm == DatasetRealm.PUBLIC and DATALAD_GITHUB_EXPORTS_ENABLED:
-        # Create Github sibling only if GitHub is enabled
-        github_sibling(ds, dataset, siblings)
-        publish_target(ds, realm.github_remote, snapshot)
-
-    clear_dataset_cache(dataset, cookies)
-
+    realm = get_realm(ds, siblings)
+    s3_sibling(ds, siblings, realm=realm)
+    if check_should_export(ds, tags):
+        for tag in tags:
+            s3_export_successful = False
+            github_export_successful = False
+            error = None
+            try:
+                publish_target(ds, realm.s3_remote, tag)
+                gevent.sleep()
+                s3_export_successful = True
+                # Public publishes to GitHub
+                if realm == DatasetRealm.PUBLIC and DATALAD_GITHUB_EXPORTS_ENABLED:
+                    github_sibling(ds, dataset_id, siblings)
+                    publish_target(ds, realm.github_remote, tag)
+                    gevent.sleep()
+                    github_export_successful = True
+            except Exception as err:
+                error = err
+            finally:
+                if esLogger:
+                    esLogger.log(tag, s3_export_successful, github_export_successful, error)
+        clear_dataset_cache(dataset, cookies)
+
+def check_s3_has_version(dataset, tag):
+    response = dataset.repo._run_annex_command(
+        'info',
+        annex_options=[
+            '--json',
+            tag,
+        ]
+    )
+    info = json.loads(response[0])
+    return info['success']
```

Comment:
This is just checking if the tag exists locally. It needs to verify the expected remote is listed in the output as well.